cpp/benchmarks/io/json/nested_json.cpp

```diff
@@ -46,17 +126,17 @@ auto make_test_json_data(size_type string_size, rmm::cuda_stream_view stream)
                       {"a":1,"b":Infinity,"c":[null], "d": {"year":-600,"author": "Kaniyan"}},
                       {"a": 1, "b": 8.0, "d": { "author": "Jean-Jacques Rousseau"}},)";
 
-  const size_type repeat_times = string_size / input.size();
+  const cudf::size_type repeat_times = string_size / input.size();
 
   auto d_input_scalar   = cudf::make_string_scalar(input, stream);
   auto& d_string_scalar = static_cast<cudf::string_scalar&>(*d_input_scalar);
   auto d_scalar         = cudf::strings::repeat_string(d_string_scalar, repeat_times);
-  auto& d_input         = static_cast<cudf::scalar_type_t<std::string>&>(*d_scalar);
 
-  auto generated_json    = std::string(d_input);
-  generated_json.front() = '[';
-  generated_json.back()  = ']';
-  return generated_json;
+  auto data = const_cast<char*>(d_scalar->data());
+  cudaMemsetAsync(data, '[', 1, stream.value());
+  cudaMemsetAsync(data + d_scalar->size() - 1, ']', 1, stream.value());
```

Comment:
Wait, you can call `cuda*` APIs in a `cpp` file?

Reply:
Yes. [C++ runtime APIs](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html) and driver APIs can be called in any C++ files. We use runtime APIs almost always.

https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-runtime
Include `cuda.h` and  `cuda_runtime.h`, Link with libcudart for runtime for non-nvcc compilers. 
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#driver-api for driver APIs.

Reply:
My mental model is that you just cannot include/declare/call any `__device__` functions. Since any `__device__` code behind `cudaMemsetAsync` (if any) is in a separate compilation unit, it's fine to call it here. Templated stuff (e.g. Thrust) is usually a no-go, since device code in usually in the headers.
I'm sure the truth is more nuanced, this is just my functional understanding.